### PR TITLE
[Reef 353] Derive OperatorTopology from an interface IOperatorTopology

### PR DESCRIPTION
**Pull request by dkm2110**
[REEF-353]: Derive OperatorTopology from an interface IOperatorTopology
This addressed the issue by
  * Creating IOperatorTopology snd making OperatorTopology derive from it.
  * In operators like BroadcastSender, BroadcastReceiver etc, use IOperstorTopology instead of OperatorTopology.

JIRA: [REEF-353](https://issues.apache.org/jira/browse/REEF-353)

### CODE REVIEW

**Comment by dkm2110** [comment]
@markusweimer @jwang98052 

SO in this pull request there are lots of .java files coming... I tried doing diff with origin/master and only 7 files are shown to be different. Not sure whats happening.

**Comment by markusweimer** [comment]
The mirror of the apache git to here seems to be stalled. Hence, you see the diff between the last time it was mirrored and your code, which include the massive recent PR to change all the license headers. It should be resolved soon and only the 7 changed files will show up here.

**Comment by markusweimer** [comment]
The sync issues are resolved. However, the PR still shows too many changed files. @dkm2110 can you `rebase` onto current `master`?

**Comment by dkm2110** [comment]
@markusweimer @jwang98052 
Seems fine now

**Comment by markusweimer** [comment]
I'm looking at this.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+using System;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: Many of these imports aren't used. Please clean those up. ReSharper is your friend.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+using Org.Apache.REEF.Network.Group.Operators;
+using Org.Apache.REEF.Network.Group.Operators.Impl;
+using Org.Apache.REEF.Network.Group.Task.Impl;
+using Org.Apache.REEF.Network.NetworkService;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Exceptions;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Remote;
+
+namespace Org.Apache.REEF.Network.Group.Task
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: The name space and folder structure don't agree.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+{
+    /// <summary>
+    /// Contains the Operator's topology graph.
+    /// Used to send or receive messages to/from operators in the same
+    /// Communication Group.
+    /// </summary>
+    /// <typeparam name="T">The message type</typeparam>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: This is confusing as there is also an `enum` called `MessageType`. What is that used for?

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        /// <param name="type">The message type</param>
+        void SendToChildren(T message, MessageType type);
+
+        /// <summary>
+        /// Splits the list of messages up evenly and sends each sublist
+        /// to the child Tasks.
+        /// </summary>
+        /// <param name="messages">The list of messages to scatter</param>
+        /// <param name="type">The message type</param>
+        void ScatterToChildren(List<T> messages, MessageType type);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: Use `IList` or better yet `IEnumerable` if we don't need `List` semantics.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        /// <summary>
+        /// Splits the list of messages up into chunks of the specified size 
+        /// and sends each sublist to the child Tasks.
+        /// </summary>
+        /// <param name="messages">The list of messages to scatter</param>
+        /// <param name="count">The size of each sublist</param>
+        /// <param name="type">The message type</param>
+        void ScatterToChildren(List<T> messages, int count, MessageType type);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: Use `IList` or better yet `IEnumerable` if we don't need `List` semantics.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+        /// <summary>
+        /// Splits the list of messages up into chunks of the specified size 
+        /// and sends each sublist to the child Tasks in the specified order.
+        /// </summary>
+        /// <param name="messages">The list of messages to scatter</param>
+        /// <param name="order">The order to send messages</param>
+        /// <param name="type">The message type</param>
+        void ScatterToChildren(List<T> messages, List<string> order, MessageType type);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: Use `IList` or better yet `IEnumerable` if we don't need `List` semantics.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        /// Receive an incoming message from the parent Task.
+        /// </summary>
+        /// <returns>The parent Task's message</returns>
+        T ReceiveFromParent();
+
+        /// <summary>
+        /// Receive a list of incoming messages from the parent Task.
+        /// </summary>
+        /// <returns>The parent Task's list of messages</returns>
+        List<T> ReceiveListFromParent();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: Use `IList` or better yet `IEnumerable` if we don't need `List` semantics.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs [hunk 8]
```diff
@@ -64,7 +64,7 @@ public class BroadcastReceiver<T> : IBroadcastReceiver<T>
             PipelineDataConverter = dataConverter;
             _topology = topology;
 
-            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => _topology.OnNext(message));
+            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => topology.OnNext(message));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: Why this change? It seems that nothing would have to change in this line.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastSender.cs [hunk 9]
```diff
@@ -66,7 +66,7 @@ public class BroadcastSender<T> : IBroadcastSender<T>
             Version = PipelineVersion;
             PipelineDataConverter = dataConverter;
 
-            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => _topology.OnNext(message));
+            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => topology.OnNext(message));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastSender.cs]: Same as above.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ReduceReceiver.cs [hunk 10]
```diff
@@ -71,7 +71,7 @@ public class ReduceReceiver<T> : IReduceReceiver<T>
             _pipelinedReduceFunc = new PipelinedReduceFunction<T>(ReduceFunction);
             _topology = topology;
 
-            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => _topology.OnNext(message));
+            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => topology.OnNext(message));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ReduceReceiver.cs]: Same as above.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ReduceSender.cs [hunk 11]
```diff
@@ -72,7 +72,7 @@ public class ReduceSender<T> : IReduceSender<T>
             _pipelinedReduceFunc = new PipelinedReduceFunction<T>(ReduceFunction);
             _topology = topology;
 
-            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => _topology.OnNext(message));
+            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => topology.OnNext(message));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ReduceSender.cs]: Same as above.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ScatterSender.cs [hunk 12]
```diff
@@ -60,7 +60,7 @@ public class ScatterSender<T> : IScatterSender<T>
             Version = DefaultVersion;
             _topology = topology;
 
-            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => _topology.OnNext(message));
+            var msgHandler = Observer.Create<GroupCommunicationMessage>(message => topology.OnNext(message));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/ScatterSender.cs]: Same as above.

File: lang/cs/Org.Apache.REEF.Network/Group/Task/IOperatorTopology.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace Org.Apache.REEF.Network.Group.Task
+{
+    /// <summary>
+    /// Contains the Operator's topology graph.
+    /// Used to send or receive messages to/from operators in the same
+    /// Communication Group.
+    /// </summary>
+    /// <typeparam name="T">The message type</typeparam>
+    public interface IOperatorTopology<T>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Task/IOperatorTopology.cs]: Is this a duplicate? The same code is at the beginning of this PR.

**Comment by markusweimer** [comment]
@dkm2110 I'm done with my pass.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs (hunk 1 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: done

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs (hunk 2 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: fixed

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs (hunk 3 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: MessageType is not used anywhere. I am not sure why is this used. Adding shravan @shravanmn

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs (hunk 7 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/IOperatorTopology.cs]: done everywhere...in once place i have to change iList to List within the code to call GetRange function

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs (hunk 8 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: _topology is of type IOperatorTopology while topology is of type OperatorTopology which also inherits IObserver. So change is needed since IOperatorTopology does not inherit it.

**Comment by dkm2110** [comment]
@markusweimer @jwang98052 

I have done the changes...if things look fine please merge and then I will modify IWritable version

**Comment by markusweimer** [comment]
Looking into it.

**Comment by markusweimer** [comment]
The tests pass and the code looks good. Merging it.

File: lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs (hunk 8 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: Why not use IOperatorTopology in operator constructors?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network/Group/Operators/Impl/BroadcastReceiver.cs]: because it should not inherit from IObserver. Should it?
